File: java-car-data/car-data-web-ui-tests/src/test/java/com/glicerial/samples/cardata/web/uitests/page/MainPage.java
package com.glicerial.samples.cardata.web.uitests.page;

import static org.junit.Assert.assertTrue;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.openqa.selenium.support.ui.Select;

import com.glicerial.samples.cardata.web.uitests.CarDataUtility;


public class MainPage {

    private WebDriver driver;
    private CarDataUtility carDataUtility = new CarDataUtility();

    public enum CarSortBy { MAKE, OLDEST, NEWEST }

    @FindBy(linkText = "Add Car")
    private WebElement addCarLink;

    public MainPage(WebDriver webDriver) {
        this.driver = webDriver;
        PageFactory.initElements(driver, this);
    }

    public void clickAddCarLink(){
        addCarLink.click();
        System.out.println("Clicked add car link");
    }

    public List<WebElement> getCarRows() {
        WebDriverWait wait = new WebDriverWait(driver, 5);
        List<WebElement> carRows = wait.until(ExpectedConditions.presenceOfAllElementsLocatedBy(By.xpath("//table[@id='carstable']/tbody/tr")));

        return carRows;
    }

    public void navigateToCarDetailPage(Map<String, String> carMap) {
        boolean navigated;

        try {
            WebElement carRow = findCarRow(carMap);
            List<WebElement> rowColumns = carRow.findElements(By.xpath(".//td"));
            rowColumns.get(0).findElement(By.tagName("a")).click();
            navigated = true;
            System.out.println("Going to detail page of car: " + carMap.get("carString"));
        } catch (NoSuchElementException e) {
            navigated = false;
        }

        assertTrue(navigated);
    }

    public WebElement findCarRow(Map<String, String> carMap) {
        System.out.println("Finding car row...");
        Boolean clickedNextPageLink = false;

        do {
            List<WebElement> carRows = getCarRows();

            for (WebElement row : carRows) {
                Map<String, String> rowCarMap = carRowToCarMap(row);

                if (carDataUtility.getCarString(rowCarMap).equals(carMap.get("carString"))) {
                    return row;
                }
          }

            clickedNextPageLink = clickNextPageLink();
        } while (clickedNextPageLink);

        return null;
    }

    public Map<String, WebElement> findMultipleCarRows(List<Map<String, String>> carMapList) {
        System.out.println("Finding multiple car rows...");
        Boolean clickedNextPageLink = false;
        int carMapListSize = carMapList.size();
        int foundCount = 0;

        Map<String, WebElement> foundCarRows = new HashMap<String, WebElement>();

        for (Map<String, String> carMap: carMapList) {
            foundCarRows.put(carMap.get("carString"), null);
        }

        do {
            List<WebElement> carRows = getCarRows();

            for (WebElement row : carRows) {
                Map<String, String> rowCarMap = carRowToCarMap(row);
                String rowCarString = carDataUtility.getCarString(rowCarMap) ;

                if (foundCarRows.containsKey(rowCarString)) {
                    foundCarRows.put(rowCarString, row);
                    System.out.println("Found car: " + rowCarString);
                    foundCount++;

                    if (foundCount == carMapListSize) {
                        return foundCarRows;
                    }
                }
            }

            clickedNextPageLink = clickNextPageLink();
        } while (clickedNextPageLink);

        return foundCarRows;
    }

    private boolean clickNextPageLink() {
        try {
            WebElement nextPageLink = driver.findElement(By.linkText("[Next]"));
            nextPageLink.click();
            System.out.println("Clicked next page link");
        } catch (NoSuchElementException ex) {
            return false;
        }

        return true;
    }

    public void selectCarSortBy(CarSortBy carSortBy) {
        WebDriverWait wait = new WebDriverWait(driver, 5);
        Select carSortBySelect = new Select(wait.until(ExpectedConditions.presenceOfElementLocated(By.id("carsortby"))));

        switch(carSortBy) {
            case MAKE:
                carSortBySelect.selectByValue("make");
                break;
            case OLDEST:
                carSortBySelect.selectByValue("oldest");
                break;
            case NEWEST:
                carSortBySelect.selectByValue("newest");
                break;
        }
    }

    public Map<String, String> carRowToCarMap(WebElement row) {
        List<WebElement> rowColumns = row.findElements(By.xpath(".//td"));
        Map<String, String> carMap = new HashMap<String, String>();

        carMap.put("year", rowColumns.get(0).getText());
        carMap.put("make", rowColumns.get(1).getText());
        carMap.put("model", rowColumns.get(2).getText());
        carMap.put("trimLevels", rowColumns.get(3).getText().replaceAll(", ", "\n"));

        return carMap;
    }
}

File: python-selenium/pages/main_page.py
from enum import Enum
from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.ui import Select
from selenium.webdriver.support.ui import WebDriverWait

from car_data_utility import get_car_string
from logging_utility import get_logger
from .locators import MainPageLocators
from .page import BasePage

""" Logging setup """
LOG = get_logger(__name__)

class MainPage(BasePage):
    class CarSortBy(Enum):
        MAKE = 1
        NEWEST = 2
        OLDEST = 3

    def click_add_car_link(self):
        add_car_link = self.driver.find_element(*MainPageLocators.ADD_CARS_LINK)
        add_car_link.click()

    def get_car_rows(self):
        wait = WebDriverWait(self.driver, 5)
        car_rows = wait.until(expected_conditions.presence_of_all_elements_located(MainPageLocators.CAR_ROWS))

        return car_rows

    def navigate_to_car_detail_page(self, car):
        car_row = self.find_car_row(car)
        row_columns = car_row.find_elements(By.XPATH, ".//td")
        row_columns[0].find_element(By.TAG_NAME, "a").click()

    def find_car_row(self, car):
        LOG.info("Finding car row...")

        while True:
            car_rows = self.get_car_rows()

            for row in car_rows:
                row_car = self.car_row_to_car_dict(row)
                row_car_string = get_car_string(row_car)

                if row_car_string == car['car_string']:
                    LOG.info("Found car: {}".format(row_car_string))
                    return row

            clicked_next_page_link = self.click_next_page_link()

            if not clicked_next_page_link:
                break

        return None

    def find_multiple_car_rows(self, car_list):
        LOG.info("Finding multiple car rows...")
        found_count = 0
        found_car_rows = {}

        for car in car_list:
            found_car_rows[car['car_string']] = None

        while True:
            car_rows = self.get_car_rows()

            for row in car_rows:
                row_car = self.car_row_to_car_dict(row)
                row_car_string = get_car_string(row_car)

                if row_car_string in found_car_rows:
                    LOG.info("Found car: {}".format(row_car_string))
                    found_car_rows[row_car_string] = row
                    found_count += 1

                    if found_count == len(car_list):
                        return found_car_rows


            clicked_next_page_link = self.click_next_page_link()

            if not clicked_next_page_link:
                break

        return found_car_rows

    def click_next_page_link(self):
        try:
            next_page_link = self.driver.find_element(By.LINK_TEXT, "[Next]")
            next_page_link.click()
            LOG.info("Clicked next page link")
        except NoSuchElementException as e:
            return False

        return True

    def select_car_sort_by(self, car_sort_by):
        select = Select(self.driver.find_element(*MainPageLocators.SORT_BY_DROP_DOWN))

        if car_sort_by == self.CarSortBy.MAKE:
            select.select_by_value("make")
        elif car_sort_by == self.CarSortBy.NEWEST:
            select.select_by_value("newest")
        elif car_sort_by == self.CarSortBy.OLDEST:
            select.select_by_value("oldest")

    def car_row_to_car_dict(self, row):
        td_elements = row.find_elements(By.XPATH, ".//td")

        return {
            'year': td_elements[0].text,
            'make': td_elements[1].text,
            'model': td_elements[2].text,
            'trim_levels': td_elements[3].text.split(", ")
        }
